### ISSUE TITLE
Empty tuple not inferred from rest parameters of a generic array type

### ISSUE THREAD

**Issue by fasttime** [microsoft/TypeScript#38770]
**TypeScript Version:**  3.9.3

**Search Terms:** generic rest parameters empty tuple

**Code**

```ts
declare function requireAtLeast1Arg<T extends any[]>(...args: T extends [] ? never : T);

requireAtLeast1Arg();
```

**Expected behavior:** The code above should trigger the infamous compiler error 2345 as it used to do in TypeScript <= 3.8.

**Actual behavior:** The code compiles in current TypeScript 3.9.3.

I couldn't find a release note about the new behavior, so I'm assuming this was a kind of unintended breaking change.

**Playground Link:**

* [Using TypeScript 3.8.3](https://www.typescriptlang.org/play?noImplicitAny=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&esModuleInterop=false&declaration=false&ts=3.8.3#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwB4AVeEADwxFWAGd4pUBPAbQF0A+ACgDp-YAczoAueGUrVaDDvAD88VCABuIGPDEkAlAG4AUHsSFuuvfHPwgA)
* [Using TypeScript 3.9.2](https://www.typescriptlang.org/play?noImplicitAny=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&esModuleInterop=false&declaration=false&ts=3.9.2#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwB4AVeEADwxFWAGd4pUBPAbQF0A+ACgDp-YAczoAueGUrVaDDvAD88VCABuIGPDEkAlAG4AUHsSFuuvfHPwgA)

**Comment by ahejlsberg** [comment]
This is an intended change, see #35438. A better way to require at least one argument is

```ts
declare function requireAtLeast1Arg<T extends [any, ...any[]]>(...args: T): void;

requireAtLeast1Arg();  // Error
requireAtLeast1Arg(42);
requireAtLeast1Arg(42, 'hello');
```

This also produces a much better error message.

**Comment by fasttime** [comment]
I see, thanks. Missed that change.
